net: lib: lwm2m: add lpwan object

marco85mars20 opened the pull request:
Add support for Low Power Wide Area Network (LPWAN) Object (ID 3412)

Signed-off-by: Marco Argiolas <marco.argiolas@ftpsolutions.com.au>

marco85mars20 committed:
net: lib: lwm2m: add uCIFI LPWAN object

Add support for Low Power Wide Area Network (LPWAN) Object (ID 3412)

Signed-off-by: Marco Argiolas <marco.argiolas@ftpsolutions.com.au>

rlubos reviewed (commented):
Looks pretty good, just some remarks regarding naming.

SeppoTakalo commented on subsys/net/lib/lwm2m/Kconfig.ucifi:
```diff
@@ -25,4 +25,87 @@ config LWM2M_UCIFI_BATTERY_INSTANCE_COUNT
+menuconfig LWM2M_UCIFI_LPWAN
+	bool "uCIFI LPWAN (3412) support"
+	help
+	  Include support for LWM2M Connectivity Monitoring Object (ID 4)
```
Mismatch of help text and object name. This is not a Connectivity Monitoring object.

marco85mars20 replied:
It has now been fixed.
Thanks

marco85mars20 commented:
@rlubos 
I believe I have applied all your suggested changes. Thanks for your review

SeppoTakalo commented on subsys/net/lib/lwm2m/Kconfig.ucifi:
```diff
@@ -25,4 +25,88 @@ config LWM2M_UCIFI_BATTERY_INSTANCE_COUNT
+
+config LWM2M_UCIFI_LPWAN_INSTANCE_COUNT
+	int "Maximum # of uCIFI LPWAN object instances"
+	depends on LWM2M_UCIFI_LPWAN
```
```suggestion
```

All these configuration parameters are already under `if LWM2M_UCIFI_LPWAN` so they don't need separate "depends on` clause.

marco85mars20 replied:
This has been fixed too.
Thanks

SeppoTakalo reviewed (changes requested):
Looks good.

Just drop those extra `depends on` clauses.

rlubos reviewed (commented):
Thanks @marco85mars20, look good to me.

One final request from my side - please squash the two commits, there's no need to add files in one commit only to remove/rename them in the next one. We strive to keep the commit history clean, thanks!

marco85mars20 commented:
> Thanks @marco85mars20, look good to me.
> 
> One final request from my side - please squash the two commits, there's no need to add files in one commit only to remove/rename them in the next one. We strive to keep the commit history clean, thanks!

Sure, I will.
I thought PRs are squashed by github when merging.

nixward commented on subsys/net/lib/lwm2m/Kconfig.ucifi:
```diff
@@ -25,4 +25,80 @@ config LWM2M_UCIFI_BATTERY_INSTANCE_COUNT
+	default 0
+	range 0 20
+	help
+	  This value sets the maximum number of IPv4 address assigned to a
+	  device that can be stored in this object.
```
```suggestion
	  This value sets the maximum number of IPv4 addresses assigned to a
	  device that can be stored in this object.
```

nixward commented on subsys/net/lib/lwm2m/Kconfig.ucifi:
```diff
@@ -25,4 +25,80 @@ config LWM2M_UCIFI_BATTERY_INSTANCE_COUNT
+	  stored in this object.
+
+config LWM2M_UCIFI_LPWAN_PEER_ADDRESS_MAX
+	int "Maximum # of addresses of peer nodes"
```
```suggestion
	int "Maximum # of peer node addresses"
```

nixward commented on subsys/net/lib/lwm2m/Kconfig.ucifi:
```diff
@@ -25,4 +25,80 @@ config LWM2M_UCIFI_BATTERY_INSTANCE_COUNT
+	  that can be stored in this object.
+
+config LWM2M_UCIFI_LPWAN_MULTICAST_GRP_ADDRESS_MAX
+	int "Maximum # of multicast addresses"
+	default 0
+	range 0 20
+	help
+	  This value sets the maximum number of group addresses (from which the
+	  device should accept incoming messages and/or commands) that can be
+	  stored in this object.
```
```suggestion
	int "Maximum # of multicast group addresses"
	default 0
	range 0 20
	help
	  This value sets the maximum number of multicast group addresses (from
	  which the device should accept incoming messages and/or commands)
	  that can be stored in this object.
```

nixward commented on subsys/net/lib/lwm2m/ucifi_lpwan.c:
```diff
@@ -0,0 +1,183 @@
+	}
+
+	/* Set default values */
+	(void)memset(res[index], 0, sizeof(res[index][0]) * ARRAY_SIZE(res[index]));
+	(void)memset(mac[index], 0, MAC_ADDRESS_LEN);
```
```suggestion
	(void)memset(res[index], 0, sizeof(res[index]));
	mac[index][0] = '\0';
```

marco85mars20 commented:
Thanks @nixward for your feedbacks, I have applied your suggestions.
I have also done the following:
- added `CONFIG_LWM2M_UCIFI_SUPPORT` in `tests/net/all/prj.conf`
- fixed a couple of bugs related to renamed macros

nixward commented on subsys/net/lib/lwm2m/ucifi_lpwan.c:
```diff
@@ -0,0 +1,181 @@
+
+	/* Check that there is no other instance with this ID */
+	for (index = 0; index < MAX_INSTANCE_COUNT; index++) {
+		if (inst[index].obj && inst[index].obj_inst_id == obj_inst_id) {
```
```suggestion
		if ((inst[index].obj != NULL) && inst[index].obj_inst_id == obj_inst_id) {
```

marco85mars20 replied:
I have applied this change too.
